keps/sig-instrumentation/1668-trace-popagating/README.md

```diff
@@ -0,0 +1,501 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes to propagate trace context across components and across a series of related objects originating from an user request. It lays the foundation for enhancing relevant but scattered logs with the trace ID as common an identifier.
+
+## Motivation
+
+Current logging for a series of related messages lacks common identifiers that can be commonly found in other distributed systems. (such as [Global Request IDs](https://specs.openstack.org/openstack/oslo-specs/specs/pike/global-req-id.html)  in openstack) This makes debugging, auditing, reproducing problems, analyzing root cause via logs across components hard, administrators and developers have to match logs by basically using timestamps and object's name as hints which may takes a huge cost especially in a scenario with a large number of requests occur in a short period of time.
+
+### Definitions
+
+**Span**: The smallest unit of a trace.  It has a start and end time, and is attached to a single trace.
+
+**Trace**: A collection of Spans which represents a single process.
+
+**Trace Context**: A reference to a Trace that is designed to be propagated across component boundaries.  Sometimes referred to as the "Span Context".  It is can be thought of as a pointer to a parent span that child spans can be attached to.
+
+### Goals
+
+- Trace context received by the API Server as part of [API Server Tracing](https://github.com/kubernetes/enhancements/issues/647) can be propagated to kubernetes components
+- A series of related objects originating from an user request can be associated by trace ID
+
+### Non-Goals
+
+- Generate new trace context(Span)
+- Replace/change existing logging, metrics, or the events API
+- Add additional telemetry to any components which is already done by [API Server Tracing](https://github.com/kubernetes/enhancements/issues/647). 
+- Run any additional OpenTelemetry components (such as the OpenTelemetry collector, which the  [API Server Tracing](https://github.com/kubernetes/enhancements/issues/647) KEP uses)
+
+## Proposal
+
+### Architecture
+
+### Trace context propagation
+
+To link work done across components as belonging to the same action(user request), we must pass trace context across process boundaries. In traditional distributed systems, this context can be passed down through RPC metadata or HTTP headers. Kubernetes, however, due to its watch-based nature, requires us to attach trace context directly to the target object.
+
+In this proposal, we choose to propagate this trace context as object annotations called `trace.kubernetes.io/context`
```

Comment:
There's currently nothing in the Alternatives section filled out. Can you discuss the pros and cons of this proposal there as well as other possible implementations for completeness?

Reply:
2 possible alternative regarding of how to propagate the 'trace context'  we have considered are:
Alternative 1: change apiserver, put what webhook does to apiserver.
    Pros: no need to deploy the mutating admission controller, easy to use.
    Cons: it will increase code coupling in apiserver, (and this feature should not belong to apiserver's responsibility at all).

Alternative 2: change each controller-manager and client(kubectl) to do work of writing trace context into related objects.
    Pros: have no idea yet
    Cons: The code change will be too much